File: practice/algorithms/strings/super_reduced_string/SuperReducedString.java
import java.util.*;

public class SuperReducedString {

    public static void main(String[] args) {
        Scanner stdin = new Scanner(System.in);
        String s = stdin.nextLine();
        String acc = "";
        for(int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if(acc.length() > 0 && acc.charAt(acc.length() - 1) == c) {
                acc = acc.substring(0, acc.length() - 1);
            } else {
                acc += c;
            }
        }
        System.out.println(acc.isEmpty() ? "Empty String" : acc);
        stdin.close();
    }
}

File: practice/algorithms/strings/super_reduced_string/SuperReducedString.scala
import scala.io.Source

object SuperReducedString extends App {
  val line = Source.stdin.getLines().toList.head
  val reduced = line.foldLeft("")(stringReduce)
  println(if(reduced.isEmpty) "Empty String" else reduced)

  def stringReduce(acc: String, c: Char): String = {
    if(acc.length > 0 && acc.charAt(acc.length - 1) == c) acc.substring(0, acc.length - 1) else acc + c
  }
}